lib/Dialect/CGGI/Transforms/StraightLineVectorizer.cpp

```diff
@@ -0,0 +1,176 @@
+#include "include/Dialect/CGGI/Transforms/StraightLineVectorizer.h"
+
+#include "include/Dialect/CGGI/IR/CGGIAttributes.h"
+#include "include/Dialect/CGGI/IR/CGGIOps.h"
+#include "include/Dialect/LWE/IR/LWEAttributes.h"
+#include "include/Graph/Graph.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"            // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"             // from @llvm-project
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"    // from @llvm-project
+#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Visitors.h"               // from @llvm-project
+#include "mlir/include/mlir/Transforms/TopologicalSortUtils.h"  // from @llvm-project
+
+#define DEBUG_TYPE "straight-line-vectorizer"
+
+namespace mlir {
+namespace heir {
+namespace cggi {
+
+#define GEN_PASS_DEF_STRAIGHTLINEVECTORIZER
+#include "include/Dialect/CGGI/Transforms/Passes.h.inc"
+
+/// Returns true if the two operations can be combined into a single vectorized
+/// operation.
+bool areCompatible(Operation *lhs, Operation *rhs) {
+  if (lhs->getName() != rhs->getName() ||
+      lhs->getDialect() != rhs->getDialect() ||
+      lhs->getResultTypes() != rhs->getResultTypes()) {
+    return false;
+  }
+
+  return llvm::TypeSwitch<Operation *, bool>(lhs)
+      // Builtin ops
+      .Case<AndOp, OrOp, XorOp>([&](auto op) { return true; })
```

Comment:
I can't re-use the Vectorizable or Tensorizable traits themselves (they require all operands can be upgraded to vectors/tensors), but I like the idea of adding a new interface for checking for compatibility. Since this pass is specific to CGGI, I think we should do that once we have the need for vectorizing in a second dialect. I suspect that arithmetic scheme FHE vectorization will have its own challenges due to ciphertext packing, that won't be as simple as generalizing this pass to arbitrary ops.